File: app/src/main/java/ru/geekbrains/notes/view/CardUpdateFragment.java
package ru.geekbrains.notes.view;

import android.content.Context;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.DatePicker;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;

import com.google.android.material.textfield.TextInputEditText;

import java.util.Calendar;
import java.util.Date;

import ru.geekbrains.notes.MainActivity;
import ru.geekbrains.notes.R;
import ru.geekbrains.notes.data.CardData;
import ru.geekbrains.notes.observation.Publisher;

public class CardUpdateFragment extends Fragment {

    private static final String ARG_CARD_DATA = "Param_CardData";
    private CardData cardData;
    private Publisher publisher;
    private TextInputEditText title;
    private TextInputEditText description;
    private DatePicker datePicker;

    public static CardUpdateFragment newInstance(CardData cardData) {
        CardUpdateFragment fragment = new CardUpdateFragment();
        Bundle args = new Bundle();
        args.putParcelable(ARG_CARD_DATA, cardData);
        fragment.setArguments(args);
        return fragment;
    }

    public static CardUpdateFragment newInstance() {
        CardUpdateFragment fragment = new CardUpdateFragment();
        return fragment;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if(getArguments() != null) {
            this.cardData = getArguments().getParcelable(ARG_CARD_DATA);
        }
    }

    @Override
    public void onAttach(@NonNull Context context) {
        super.onAttach(context);
        this.publisher = ((MainActivity) context).getPublisher();
    }

    @Override
    public void onDetach() {
        super.onDetach();
        this.publisher = null;
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_update_card, container, false);

        initView(view);

        if (this.cardData != null)
            populateView();
        return view;
    }

    private void initView(View view) {
        this.title = view.findViewById(R.id.inputTitle);
        this.description = view.findViewById(R.id.inputDescription);
        this.datePicker = view.findViewById(R.id.inputDate);
    }

    private void populateView() {
            title.setText(cardData.getTitle());
            description.setText(cardData.getText());
            initDatePicker(cardData.getDate());
    }

    private void initDatePicker(Date date) {
        Calendar calendar = Calendar.getInstance();
        calendar.setTime(date);
        this.datePicker.init(calendar.get(Calendar.YEAR),
                calendar.get(Calendar.MONTH),
                calendar.get(Calendar.DAY_OF_MONTH),
                null);
    }

    private CardData collectCardData(){
        String title = this.title.getText().toString();
        String description = this.description.getText().toString();
        Date date = getDateFromDatePicker();
        if(cardData != null) {
            cardData.setTitle(title);
            cardData.setText(description);
            cardData.setDate(date);
            return cardData;
        }
        return new CardData(title, description, date);
    }

    private Date getDateFromDatePicker() {
        Calendar cal = Calendar.getInstance();
        cal.set(Calendar.YEAR, this.datePicker.getYear());
        cal.set(Calendar.MONTH, this.datePicker.getMonth());
        cal.set(Calendar.DAY_OF_MONTH, this.datePicker.getDayOfMonth());
        return cal.getTime();
    }

    @Override
    public void onStop() {
        super.onStop();
        cardData = collectCardData();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        publisher.notifyTask(cardData);
    }
}

File: app/src/main/java/ru/geekbrains/notes/view/CardUpdateFragment.kt
package ru.geekbrains.notes.view

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.DatePicker
import androidx.fragment.app.Fragment
import com.google.android.material.textfield.TextInputEditText
import ru.geekbrains.notes.R
import ru.geekbrains.notes.data.CardData
import ru.geekbrains.notes.presenter.MainPresenter
import java.util.*

class CardUpdateFragment(private val presenter: MainPresenter) : Fragment() {
    private var cardData = CardData()
    private lateinit var title: TextInputEditText
    private lateinit var description: TextInputEditText
    private lateinit var datePicker: DatePicker

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        if (arguments != null) {
            requireArguments().getParcelable<CardData>(ARG_CARD_DATA)?.let { cardData = it }
        }
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        val view = inflater.inflate(R.layout.fragment_update_card, container, false)
        initView(view)
        populateView()
        return view
    }

    private fun initView(view: View) {
        title = view.findViewById(R.id.inputTitle)
        description = view.findViewById(R.id.inputDescription)
        datePicker = view.findViewById(R.id.inputDate)
    }

    private fun populateView() {
        title.setText(cardData.title)
        description.setText(cardData.text)
        initDatePicker(cardData.date)
    }

    private fun initDatePicker(date: Date) {
        val calendar = Calendar.getInstance()
        calendar.time = date
        datePicker.init(
            calendar[Calendar.YEAR],
            calendar[Calendar.MONTH],
            calendar[Calendar.DAY_OF_MONTH],
            null
        )
    }

    private fun collectCardData(): CardData {
        val title = title.text.toString()
        val description = description.text.toString()
        val date = dateFromDatePicker
        cardData.title = title
        cardData.text = description
        cardData.date = date
        return cardData
    }

    private val dateFromDatePicker: Date
        get() {
            val cal = Calendar.getInstance()
            cal[Calendar.YEAR] = datePicker.year
            cal[Calendar.MONTH] = datePicker.month
            cal[Calendar.DAY_OF_MONTH] = datePicker.dayOfMonth
            return cal.time
        }

    override fun onStop() {
        super.onStop()
        cardData = collectCardData()
    }

    override fun onDestroy() {
        super.onDestroy()
        presenter.notify(cardData)
    }

    companion object {
        private const val ARG_CARD_DATA = "Param_CardData"
        fun newInstance(cardData: CardData, presenter: MainPresenter): CardUpdateFragment {
            val fragment = CardUpdateFragment(presenter)
            val args = Bundle()
            args.putParcelable(ARG_CARD_DATA, cardData)
            fragment.arguments = args
            return fragment
        }

        fun newInstance(presenter: MainPresenter): CardUpdateFragment {
            return CardUpdateFragment(presenter)
        }
    }
}
